trieste/models/gpflow/models.py

```diff
@@ -1362,3 +1368,259 @@ def covariance_between_points(
+            query_points_wo_fidelity, num_samples
+        )
+
+        for fidelity in range(self.num_fidelities):
+            if fidelity == 0:
+                fidelity_residual_sample = 0
+            else:
+                fidelity_residual_sample = self.fidelity_residual_models[fidelity].sample(
+                    query_points_wo_fidelity, num_samples
+                )
+
+            new_fidelity_signal_sample = (
+                self.rho[fidelity] * signal_sample + fidelity_residual_sample
+            )
+
+            mask = query_points_fidelity_col >= fidelity
+
+            signal_sample = tf.where(mask, new_fidelity_signal_sample, signal_sample)
+
+        return signal_sample
+
+    def predict_y(self, query_points: TensorType) -> tuple[TensorType, TensorType]:
+        """
+        Predict the marginal mean and variance at `query_points` including observation noise
+
+        :param query_points: Query points with shape [N, D+1], where the
+            final column of the final dimension contains the fidelity of the query point
+        :return: mean: The mean at query_points with shape [N, P],
+                 and var: Th variance at query_points with shape [N, P]
+        """
+
+        f_mean, f_var = self.predict(query_points)
+        query_points_fidelity_col = query_points[:, -1:]
+
+        # Get fidelity 0 observation noise
+        observation_noise = (
+            tf.ones_like(query_points_fidelity_col)
+            * self.lowest_fidelity_signal_model.get_observation_noise()
+        )
+
+        for fidelity in range(1, self.num_fidelities):
+
+            fidelity_observation_noise = (
+                self.rho[fidelity] ** 2
+            ) * observation_noise + self.fidelity_residual_models[fidelity].get_observation_noise()
+
+            mask = query_points_fidelity_col >= fidelity
+
+            observation_noise = tf.where(mask, fidelity_observation_noise, observation_noise)
+
+        return f_mean, f_var + observation_noise
+
+    def update(self, dataset: Dataset) -> None:
+        """
+        Update the models on their corresponding data. The data for each model is
+        extracted by splitting the observations in ``dataset`` by fidelity level.
+
+        :param dataset: The query points and observations for *all* the wrapped models.
+        """
+        check_and_extract_fidelity_query_points(
+            dataset.query_points, max_fidelity=self.num_fidelities - 1
+        )
+        dataset_per_fidelity = split_dataset_by_fidelity(dataset, self.num_fidelities)
+        for fidelity, dataset_for_fidelity in enumerate(dataset_per_fidelity):
+            if fidelity == 0:
+                self.lowest_fidelity_signal_model.update(dataset_for_fidelity)
+            else:
+                # Make query points but with final column corresponding to
+                # fidelity we wish to predict at
+                self.fidelity_residual_models[fidelity].update(
```

Comment:
For my understanding: I thought that the AR1 model was taking into account the prediction uncertainty of the lower layer. But this is not the case here since we only pass the residual values? Is it because I have it wrong, or because it's hard to do? In the latter case then this model isn't an AR1? which might be OK for now but we'd need to write an issue to fix it.

Reply:
looks like it's taken into account in the loss, but possibly not in predict. We could use the new heteroscedastic likelihood of gpflow maybe? (I might also have this wrong)